pkg/registry/apis/folders/register.go

```diff
@@ -256,7 +249,53 @@ func (b *FolderAPIBuilder) Validate(ctx context.Context, a admission.Attributes,
 	if !ok {
 		return fmt.Errorf("obj is not v0alpha1.Folder")
 	}
+	verb := a.GetOperation()
+
+	switch verb {
+	case admission.Create:
+		return b.validateOnCreate(ctx, id, obj)
+	case admission.Delete:
+		return b.validateOnDelete(ctx, f)
+	case admission.Update:
+		return nil
+	case admission.Connect:
+		return nil
+	}
+	return nil
+}
+
+func (b *FolderAPIBuilder) validateOnDelete(ctx context.Context, f *v0alpha1.Folder) error {
+	resp, err := b.searcher.GetStats(ctx, &resource.ResourceStatsRequest{Namespace: f.Namespace, Folder: f.Name})
+	if err != nil {
+		return err
+	}
+
+	if resp.Error != nil {
+		return fmt.Errorf("could not verify if folder is empty: %v", resp.Error)
+	}
+
+	if resp == nil || resp.Stats == nil {
+		return fmt.Errorf("could not verify if folder is empty: %v", resp.Error)
+	}
+
+	for _, v := range resp.Stats {
+		fmt.Println(v.Count)
+		if v.Count > 0 {
+			return folder.ErrFolderNotEmpty
+		}
+	}
+
+	return nil
+}
+
+func (b *FolderAPIBuilder) validateOnCreate(ctx context.Context, id string, obj runtime.Object) error {
+	for _, invalidName := range folderValidationRules.invalidNames {
+		if id == invalidName {
+			return dashboards.ErrFolderInvalidUID
+		}
+	}
 
+	f, _ := obj.(*v0alpha1.Folder)
```

Comment:
Blocking: this would panic if the type assertion fails. We need to do something like that:
```suggestion
	f, ok := obj.(*v0alpha1.Folder)
	if !ok {
	  return fmt.Errorf("type assertion failed expected *v0alpha1.Folder")
	}
```

Reply:
We are covering that case on the [caller side](https://github.com/grafana/grafana/pull/97708/files#diff-642ac82aeda3ecff99684b9d8cae8598b7c4cfa72dfcac489824a4c835e6f8ceL253-L255) and considering this is a private fn, I don't think we need to check it again? I considered passing `f` or accessing the title from `obj`, but this looked like the most readable option.

Reply:
you're right :) changed